EventsByTag.md

```diff
@@ -0,0 +1,120 @@
+* Event missing: N scans from the previous event for this persistenceId/tag's offset to the current offset.
+* Omitted events: No, stream will fail if either of the above two scenarios don't find the missing event.
+
+#### Current query / Offset
+
+* First event missing: Back track to Offset. N scans of each bucket then assume it is the first event from this offset.
+* Event missing: N scans from the previous event for this persistenceId/tag's offset to the current offset.
+* Omitted events: Yes. If first event missing doesn't find an event that should be there.
+
+
+#### Live query / No Offset
+
+* First event missing: Back track to `first-time-bucket`. N scans of each bucket.
+* Event missing: N scans from the previous event for this persistenceId/tag's offset to the current offset.
+* Omitted events: No, stream will fail if either of the above two scenarios don't find the missing event.
+
+#### Live query / Offset
+
+* First event missing: Back track to Offset. N scans of each bucket then assume it is the first event from this offset.
+* Event missing: N scans from the previous event for this persistenceId/tag's offset to the current offset.
+* Omitted events: Yes. If first event missing doesn't find an event that should be there.
+
+## Implementation
+
+A `TagWriter` actor is created, on demand, per tag that batches (unlogged batch grouped by the partition key)
+up writes to a separate table partitioned by tag name and a time bucket.
+
+When a message is persisted, if it has tags, it is sent to each of the `TagWriter`s asynchronously.
+
+This allows us to be more efficient and thus support more tags than the old version (limited to 3). No limit is 
+imposed with new implementation but it is advised to keep it small i.e. less than 10.
+
+
+### Crash recovery
+
+`TagWriter` maintains a separate table with how far it has gotten up to
+for each persistenceId.
+
+```
+CREATE TABLE akka.tag_views_progress (                              
+    peristence_id text,           
+    tag text,                     
+    sequence_nr bigint,           
+    PRIMARY KEY (peristence_id, tag)                                
+)
+```
+
+Part of recovery will then check that all events have been written by the tag writer.
+
+On the happy path this table will be written to after the tag_view write is complete
+so it could be lost in a crash. In that event recovery will duplicate writes to the 
+tag_views table which is fine as they will be upserts.
+
+TODO:
+What happens if a persistenceId is not uses for a long time after a crash? eventsByTag queries 
+would remain out of date for a long time.
+
+### Deletion of events
+
+*Deletes are not propagated to the events by tag table.*
+
+It is assumed that the eventsByTag query is used to populate a separate read view or
+events are never deleted. An optional TTL can be set on the tag_view table to clean up
+events assuming they have been written to another read view database, otherwise they are
+kept forever.
```

Comment:
would it be possible to implement this feature at a later time? I suppose it somehow would have to iterate through all persistenceIds, finding any that have deleted events, and then go in and remove them from the materialized tables' time windows... that would be a lot of scanning.

Reply:
It would be hard to retrospectively do it without scanning most of the `tag_views` table as it isn't indexed by persitence id and we don't know which tags a deleted event had once it is gone from the messages table.